lib/matplotlib/dates.py

```diff
@@ -740,12 +746,237 @@ def __call__(self, x, pos=0):
+        elif formats:
+            # use the users formats for the zero tick formats
+            self.zero_formats = [''] + self.formats[:-1]
+        else:
+            # make the defaults a bit nicer:
+            self.zero_formats = [''] + self.formats[:-1]
+            self.zero_formats[3] = '%b-%d'
+
+        if offset_formats:
+            if len(offset_formats) != 6:
+                raise ValueError('offsetfmts argument must be a list of '
+                                 '6 format strings (or None)')
+            self.offset_formats = offset_formats
+        else:
+            self.offset_formats = ['',
+                                   '%Y',
+                                   '%Y-%b',
+                                   '%Y-%b-%d',
+                                   '%Y-%b-%d',
+                                   '%Y-%b-%d %H:%M']
+        self.offset_string = ''
+        self._formatter = DateFormatter(self.defaultfmt, self._tz)
+        self.show_offset = show_offset
+
+    def __call__(self, x, pos=None):
+        if hasattr(self._locator, '_get_unit'):
+            locator_unit_scale = float(self._locator._get_unit())
+        else:
+            locator_unit_scale = 1.0
+        ticks = self._locator()
+        if pos is not None:
+            if not np.array_equal(ticks, self._oldticks):
+
+                offset_fmt = ''
+                fmt = self.defaultfmt
+                self._formatter = DateFormatter(fmt, self._tz)
+                tickdate = np.array([num2date(tick).timetuple()[:6]
+                                     for tick in ticks])
+                tickdatetime = [num2date(tick) for tick in ticks]
+
+                # basic algorithm:
+                # 1) only display a part of the date if it changes over the
+                #    ticks.
+                # 2) don't display the smaller part of the date if:
+                #    it is always the same or if it is the start of the
+                #    year, month, day etc.
+                # fmt for most ticks at this level
+                fmts = self.formats
+                # format beginnings of days, months, years, etc...
+                zerofmts = self.zero_formats
+                # offset fmt are for the offset in the upper left of the
+                # or lower right of the axis.
+                offsetfmts = self.offset_formats
+
+                # determine the level we will label at:
+                # mostly 0: years,  1: months,  2: days,
+                # 3: hours, 4: minutes, 5: seconds, 6: microseconds
+                for level in range(5, -1, -1):
+                    if len(np.unique(tickdate[:, level])) > 1:
+                        break
+
+                # level is the basic level we will label at.
+                # now loop through and decide the actual ticklabels
+                zerovals = [0, 1, 1, 0, 0, 0, 0]
+                ticknew = []
+                for nn in range(len(tickdate)):
```

Comment:
These could be `zip`d, no?

Reply:
Not sure what you mean...

Reply:
I mean, you don't seem to need `nn`: `for something, something_else in zip(tickdate, tickdatetime):` (I'm just not sure what to name those two.)

Reply:
Gotcha, but I find that actually makes me have to read the code more than what this does.  However, to make things better, I got rid of the `append` and pre-allocate the list of string first (and hence need the index to know which string to populate).